README
Welcome to BatCop v.01
----------------------

BatCop is a tool derived from Intel PowerTOP. PowerTop
was designed to spot applications that were consuming
too much power. Although initially designed to be a
diagnosis tool, we decided PowerTOP can be used for much
more than that. We thus present, BatCop, a battery attack
detection system. The equation leading to BatCop is as follows:

PowerTOP - UI - suggestions + Statistical Analysis Engine + Training Mechanism = BatCop

We leverage upon the diagnosis capabilities of PowerTOP, 
perform our own statistical analysis of the values reported by
PowerTOP's IRQ monitoring system, and detect anomalies in the
power consumption profiles of each application.

To train BatCop:

$: sudo batcop --mode=0

Leave this running for a while to let BatCop analyse the
current profile of the running system. The more data it collects,
the stronger it's guess will be when it attempts to monitor
the system later on for anomalies. Terminating the program
will lead to a trace file being generated. This can be used
as an input for the monitor mode operation.

To monitor the system using BatCop:

$: sudo batcop --mode=1 --file=<dump_file>

To do both, (not stable as far as verdicts are concerned):

$: sudo batcop --mode=2


Credits
-------
Lalith Suresh
Wasif Malik
Paulo Ricardo


debian/changelog
powertop (1.11+git20091208-0ubuntu1) lucid; urgency=low

  * Update to latest version in git on 20091208
     + Brings us info about disk wakeups and HDA audio state
  * Drop following patches since they are now upstream
     + patches/increase-buffer-for-kernel-config.dpatch
     + patches/remove-bashisms.dpatch
  * Fixed debuild error due to manpages containing powertop.1 instead of .8

 -- Amit Kucheria <amit.kucheria@canonical.com>  Tue, 08 Dec 2009 17:02:56 +0200

powertop (1.11-1ubuntu1) karmic; urgency=low

  * debian/patches/increase-buffer-for-kernel-config.dpatch:
    - Fix a segfault when kernel config is longer than 5000 lines
      (LP: #428254)

 -- Simon Olofsson <simon@olofsson.de>  Fri, 18 Sep 2009 14:40:37 +0200

powertop (1.11-1) unstable; urgency=low

  * New upstream release
     + reports now correct wattage (at least for me) (Closes: #497929)
  * Add patch in order to remove bashism in output (Closes: #509811)
  * Removed patches as they are now applied upstream:
     + patches/fix-eof.dpatch
     + patches/add-support-for-charge_now.dpatch
     + patches/fix-suggest-usb-autosuspend.dpatch
     + patches/fix-hal-disable-polling.dpatch
 

 -- Patrick Winnertz <winnie@debian.org>  Tue, 13 Jan 2009 00:02:01 +0100

powertop (1.10-2) unstable; urgency=high

  * Support batteries that do not report energy_now, by reading change_now.
    (Closes: #486873, #495638, #495348)

 -- Patrick Winnertz <winnie@debian.org>  Mon, 01 Sep 2008 20:31:51 +0200

powertop (1.10-1) unstable; urgency=low

  * New upstream version (Closes: #486510)
    - changed from proc to sys interface (Closes: #462796, #483603)
  * Build a PO template on package build (Closes: #471852)
  * spell Linux correctly: Linux (control file)
  * powertop now quits if it get a EOF (Closes: 458984)
  * Updated watch file (Closes: #486509)
  * Removed usb patch, source has changed completely. hopefully it was fixed
    by upstream.  

 -- Patrick Winnertz <winnie@debian.org>  Mon, 16 Jun 2008 23:55:10 +0200

powertop (1.9-2) unstable; urgency=low

  * Changed my maintainer address
  * Bumped Standards-Version to 3.7.3.
  * Instead of using /dev/scd0 as disk drive use /dev/cdrom for
    hal-disable-polling (Created patch fix-hal-disable-polling.dpatch) 
    (Closes: #443594)
  * This time fix really the usb autosuspend issue. (Closes: #444817)
  * Add a Homepage field to control.

 -- Patrick Winnertz <winnie@debian.org>  Sun, 16 Dec 2007 23:33:32 +0100

powertop (1.9-1) unstable; urgency=low

  * Packaging new upstream release. (Closes: #451370)
  * Fix the wrong implementation of usb autosuspend (Closes: #444817)

 -- Patrick Winnertz <patrick.winnertz@skolelinux.org>  Thu, 22 Nov 2007 13:17:38 +0100

powertop (1.8-1) unstable; urgency=low

  * Packaging new upstream release. 
  * Correct the dependency to laptop-mode-tools. (Closes: #438616)
  * Correct the command line syntax of hal-disable-polling (Closes: #435882)
  * Allow powertop to build on all archs since some informations are still
    there (wakeup counts).

 -- Patrick Winnertz <patrick.winnertz@skolelinux.org>  Mon, 20 Aug 2007 22:23:53 +0200

powertop (1.7~svn-r227-3) unstable; urgency=low

  * debian/control:
   + Changed the `arch" field to: `i386 amd64 ia64'. (Closes: #435972)
  * debian/copyright:
   + Fixed the license related issue, updated to just GPL2 including the
   license address to `/usr/share/common-licenses/GPL-2'. (Closes: #436367)

 -- Jose Luis Rivas <ghostbar38@gmail.com>  Thu, 16 Aug 2007 23:01:09 -0400

powertop (1.7~svn-r227-2) unstable; urgency=low

  * Thanks to Krzysztof Burghardt for the new package release, but
    remove him as co-maintainer, as he is none after all

 -- Patrick Winnertz <patrick.winnertz@skolelinux.org>  Fri, 13 Jul 2007 20:39:21 +0200

powertop (1.7~svn-r227-1) unstable; urgency=low

  * New co-maintainer
  * New upstream (closes: #429305, #430035, #427345)
  * Moved binary to /usr/sbin as it need to be run by root (closes: #427548)
  * debian/copyright:
   + changed a bit to fulfill standard look

 -- Krzysztof Burghardt <krzysztof@burghardt.pl>  Sat, 23 Jun 2007 20:30:02 +0200

powertop (1.5-2) unstable; urgency=low

  * debian/control:
   + Two leading spaces for Homepage (Closes: #426606)
  * debian/changelog:
   + Fixed typo on "Cooperation" now "Corporation" (Closes: #426605)

 -- Jose Luis Rivas <ghostbar38@gmail.com>  Wed, 30 May 2007 08:46:08 -0400

powertop (1.5-1) unstable; urgency=low

  * New upstream release available (Closes: #426507)
  * Adding Build-Depends: libncursesw5-dev since powertop fails to build without it.
  * Modified debian/copyright:
    - added (c) 2007 Intel Corporation
    - powertop is licensed under GPL v2 (added v2)

 -- Patrick Winnertz <patrick.winnertz@skolelinux.org>  Tue, 29 May 2007 13:43:06 +0200

powertop (1.4-2) unstable; urgency=low

  [ Jose Luis Rivas ]
  * debian/control:
   + Fixed the Homepage field (Closes: #426205)
   + Modified the uploader name to "Jose Luis Rivas"

  [ Patrick Winnertz ]
  * Fixed FTBFS, by depending on libncurses5-dev (Closes: #426235)
  * Uses $(CURDIR) instead of `pwd` in rules.
  * Upload sponsored by Petter Reinholdtsen 

 -- Patrick Winnertz <patrick.winnertz@skolelinux.org>  Sun, 27 May 2007 17:59:38 +0200

powertop (1.4-1) unstable; urgency=low

  * New upstream release
  * Now powertop uses the full width of a terminal. (Fixed by upstream)
    (Closes: #423735)
  * Changed the destdir in debian/rules so that the locale files are installed correctly
    to debian/powertop/usr/share/locale/:
      - DESTDIR=debian/powertop  --> DESTDIR=`pwd`/debian/powertop
  * Upload sponsored by Petter Reinholdtsen

 -- Patrick Winnertz <patrick.winnertz@skolelinux.org>  Sat, 26 May 2007 17:24:27 +0200

powertop (1.3-1) unstable; urgency=low

  * New upstream release (Closes: #425498)
  * Suggests cpufrequtils (Closes: #423986)
  * Clean up debian/rules:
      - removes simple-patchsys.mk since it not used

perf.c
/*
 * Copyright 2009, Intel Corporation
 *
 * This file is part of PowerTOP
 *
 * Portions borrowed from the "perf" tool (C) Ingo Molnar and others
 *
 * This program file is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program in a file named COPYING; if not, write to the
 * Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA
 *
 * Authors:
 * 	Arjan van de Ven <arjan@linux.intel.com>
 */

#define _GNU_SOURCE        /* or _BSD_SOURCE or _SVID_SOURCE */
#include <unistd.h>
#include <sys/syscall.h>   /* For SYS_xxx definitions */
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ioctl.h>

#include <fcntl.h>

#include "perf_event.h"

#include "batcop.h"


/* some people have stale headers */
#ifndef __NR_perf_event_open
#ifdef __i386__
#define __NR_perf_event_open	336
#endif
#if __x86_64__
#define __NR_perf_event_open	298
#endif
#endif

static int perf_fd = -1;
static void * perf_mmap;
static void * data_mmap;

static struct perf_event_mmap_page *pc;

static inline int
sys_perf_event_open(struct perf_event_attr *attr,
                      pid_t pid, int cpu, int group_fd,
                      unsigned long flags)
{
	attr->size = sizeof(*attr);
	return syscall(__NR_perf_event_open, attr, pid, cpu,
			group_fd, flags);
}

static int this_trace = 0;

static int get_trace_type(void)
{
	FILE *file;
	char line[4096];
	file = fopen("/sys/kernel/debug/tracing/events/vfs/dirty_inode/id", "r");
	if (!file)
		return 0;

	if (fgets(line, 4095, file) == NULL)
		return 0;
	
	this_trace = strtoull(line, NULL, 10);
	fclose(file);
	return this_trace;
}

static void create_perf_event(void)
{
	struct perf_event_attr attr;
	int ret;

	struct {
		__u64 count;
		__u64 time_enabled;
		__u64 time_running;
		__u64 id;
	} read_data;

	memset(&attr, 0, sizeof(attr));

	attr.read_format	= PERF_FORMAT_TOTAL_TIME_ENABLED |
				  PERF_FORMAT_TOTAL_TIME_RUNNING |
				  PERF_FORMAT_ID;

	attr.sample_freq 	= 0;
	attr.sample_period	= 1;
	attr.sample_type	|= PERF_SAMPLE_RAW;

	attr.mmap		= 1;
	attr.comm		= 1;
	attr.inherit		= 0;
	attr.disabled		= 1;

	attr.type		= PERF_TYPE_TRACEPOINT;
	attr.config		= get_trace_type();

	if (attr.config <= 0)
		return;

	perf_fd = sys_perf_event_open(&attr, -1, 0, -1, 0);

	if (perf_fd < 0) {
		fprintf(stderr, "Perf syscall failed: %i / %i (%s)\n", perf_fd, errno, strerror(errno));
		return;
	}
	if (read(perf_fd, &read_data, sizeof(read_data)) == -1) {
		perror("Unable to read perf file descriptor\n");
		exit(-1);
	}

	fcntl(perf_fd, F_SETFL, O_NONBLOCK);

	perf_mmap = mmap(NULL, (128+1)*getpagesize(),
				PROT_READ | PROT_WRITE, MAP_SHARED, perf_fd, 0);
	if (perf_mmap == MAP_FAILED) {
		fprintf(stderr, "failed to mmap with %d (%s)\n", errno, strerror(errno));
		return;
	}

	ret = ioctl(perf_fd, PERF_EVENT_IOC_ENABLE);

	if (ret < 0)
		fprintf(stderr, "failed to enable perf \n");

	pc = perf_mmap;
	data_mmap = perf_mmap + getpagesize();


}


void start_data_dirty_capture(void)
{
	create_perf_event();
}

void end_data_dirty_capture(void)
{
	if (perf_fd >= 0)
		close(perf_fd);
	perf_fd = -1;
}

struct trace_entry {
	__u32			size;
	unsigned short		type;
	unsigned char		flags;
	unsigned char		preempt_count;
	int			pid;
	int			tgid;
};

#define TASK_COMM_LEN 16
struct dirty_inode { 
	char comm[TASK_COMM_LEN];
	int  pid;
	char dev[16];
	char file[32];
};


struct sample_event{
	struct perf_event_header        header;
	struct trace_entry		trace;
	struct dirty_inode		inode;
};


static void parse_event(void *ptr, int verbose)
{
	char line[8192];
	char pid[14];
	int suggested = 0;
	struct sample_event *event = ptr;
	if (event->trace.type != this_trace)
		return;

	if (event->trace.size < sizeof(struct dirty_inode)) 
		return;

	if (event->inode.pid == 0)
		return;

	if (strcmp(event->inode.comm, "powertop") == 0)
		return;
	/*
	 * btrfs kernel threads are internal and only
	 * do IO on behalf of others that also got recorded
	 */

	if (strcmp(event->inode.comm, "btrfs-") == 0)
		return;
	/*
	 * don't record "IO" to tmpfs or /proc
	 */
	if (strcmp(event->inode.dev, "tmpfs") == 0)
		return;
	if (strcmp(event->inode.dev, "proc") == 0)
		return;
	if (strcmp(event->inode.dev, "pipefs") == 0)
		return;
	if (strcmp(event->inode.dev, "anon_inodefs") == 0)
		return;

	sprintf(pid, "%i", event->inode.pid);
	sprintf(line, "%s", event->inode.comm);
	push_line_pid(line, 0, 1, pid);

	if (!suggested && strcmp(event->inode.file, "?")) {
		suggested = 1;
		sprintf(line,_("The program '%s' is writing to file '%s' on /dev/%s.\nThis prevents the disk from going to powersave mode."),
			event->inode.comm, event->inode.file, event->inode.dev);
	}
	if (verbose) 
		printf(_("The application '%s' is writing to file '%s' on /dev/%s\n"),
			event->inode.comm, event->inode.file, event->inode.dev);

}

void parse_data_dirty_buffer(void)
{
	struct perf_event_header *header;
	int i = 0;

	if (perf_fd < 0)
		return;

	while (pc->data_tail != pc->data_head && i++ < 5000) {
		while (pc->data_tail >= 128U * getpagesize())
			pc->data_tail -= 128 * getpagesize();

		header = data_mmap + pc->data_tail;

		if (header->size == 0)
			break;

		pc->data_tail += header->size;

		while (pc->data_tail >= 128U * getpagesize())
			pc->data_tail -= 128 * getpagesize();

  	parse_event(header, 0);
	}
	pc->data_tail = pc->data_head;
}
